sysl2/sysl/ints.go

```
@@ -0,0 +1,219 @@
+	for _, stmt := range stmts {
+		switch t := stmt.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			targetName := getAppName(t.Call.GetTarget())
+			if ok := v.addAppDependency(appname, epname, targetName, t.Call.Endpoint, stmt); !ok {
+				continue
+			}
+			v.addEndpoint(targetName, t.Call.Endpoint)
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			v.handleStatement(appname, epname, t.Cond.GetStmt())
+		case *sysl.Statement_Loop:
+			v.handleStatement(appname, epname, t.Loop.GetStmt())
+		case *sysl.Statement_LoopN:
+			v.handleStatement(appname, epname, t.LoopN.GetStmt())
+		case *sysl.Statement_Foreach:
+			v.handleStatement(appname, epname, t.Foreach.GetStmt())
+		case *sysl.Statement_Group:
+			v.handleStatement(appname, epname, t.Group.GetStmt())
+		case *sysl.Statement_Alt:
+			for _, choice := range t.Alt.GetChoice() {
+				v.handleStatement(appname, epname, choice.GetStmt())
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func (v *IntegrationParam) addAppDependency(sourceApp, sourceEndpt, targetApp, targetEndpt string, stmt *sysl.Statement) bool {
+	if len(v.apps) == 0 {
+		v.apps = v.apps.Union(v.highlights)
+	}
+	apps := MakeStrSet(sourceApp, targetApp)
+	if len(apps.Intersection(v.excludes)) > 0 {
+		return false
+	}
+	if len(apps.Intersection(v.highlights)) > 0 {
+		if !hasPattern("human", v.m.GetApps()[sourceApp].GetAttrs()) {
+			v.apps.Insert(sourceApp)
+		}
+		if !hasPattern("human", v.m.GetApps()[targetApp].GetAttrs()) {
+			v.apps.Insert(targetApp)
+		}
+	}
+	if !(apps.IsSubSet(v.apps) || (v.apps.Contains(sourceApp) && v.passthroughs.Contains(targetApp)) || v.passthroughs.Contains(sourceApp)) {
+		return false
+	}
+	if len(MakeStrSet(sourceEndpt, targetEndpt).Intersection(v.commonEndpts)) > 0 {
+		return false
+	}
+	k := fmt.Sprintf("%s:%s:%s:%s", sourceApp, sourceEndpt, targetApp, targetEndpt)
+	if _, has := v.deps[k]; has {
+		return false
```

Again this can be moved right at the start (as 2nd check).
Fail early.

---

moved